client/sftp_client/client.py

```diff
@@ -216,28 +216,104 @@ def __init__(self, application, ssh, config):
+		self.editor = None
+		self.window = sftp_utilities.get_object('SFTPClient.window')
+		self.notebook = sftp_utilities.get_object('SFTPClient.notebook')
+		self.notebook.set_show_tabs(False)
+		self.notebook.connect('switch-page', self.signal_change_page)
+		self.editor_tab_save_button = sftp_utilities.get_object('SFTPClient.notebook.page_editor.toolbutton_save_html_file')
+		self.editor_tab_save_button.set_sensitive(False)
+		self.editor_tab_save_button.connect('clicked', self.signal_editor_save)
+		self.status_display = StatusDisplay(self.queue)
+		self.local = directory.LocalDirectory(self.application, config)
+		self.remote = directory.RemoteDirectory(self.application, config, ssh)
+		sftp_utilities.get_object('SFTPClient.notebook.page_stfp.button_upload').connect('button-press-event', lambda widget, event: self._queue_transfer_from_selection(tasks.UploadTask))
+		sftp_utilities.get_object('SFTPClient.notebook.page_stfp.button_download').connect('button-press-event', lambda widget, event: self._queue_transfer_from_selection(tasks.DownloadTask))
 		self.local.menu_item_transfer.connect('activate', lambda widget: self._queue_transfer_from_selection(tasks.UploadTask))
 		self.remote.menu_item_transfer.connect('activate', lambda widget: self._queue_transfer_from_selection(tasks.DownloadTask))
-		menu_item = self.builder.get_object('SFTPClient.notebook.page_stfp.menuitem_opts_transfer_hidden')
+		self.local.menu_item_edit.connect('activate', self.signal_edit_local_file)
+		self.remote.menu_item_edit.connect('activate', self.signal_edit_remote_file)
+		menu_item = sftp_utilities.get_object('SFTPClient.notebook.page_stfp.menuitem_opts_transfer_hidden')
 		menu_item.set_active(self.config['transfer_hidden'])
 		menu_item.connect('toggled', self.signal_toggled_config_option, 'transfer_hidden')
-		menu_item = self.builder.get_object('SFTPClient.notebook.page_stfp.menuitem_opts_show_hidden')
+		menu_item = sftp_utilities.get_object('SFTPClient.notebook.page_stfp.menuitem_opts_show_hidden')
 		menu_item.set_active(self.config['show_hidden'])
 		menu_item.connect('toggled', self.signal_toggled_config_option_show_hidden)
-		menu_item = self.builder.get_object('SFTPClient.notebook.page_stfp.menuitem_exit')
+		menu_item = sftp_utilities.get_object('SFTPClient.notebook.page_stfp.menuitem_exit')
 		menu_item.connect('activate', lambda _: self.window.destroy())
 		self.window.connect('destroy', self.signal_window_destroy)
 		self.window.show_all()
 
+	def signal_change_page(self, _, __, page_number):
+		# page_number is the page switched from
+		if page_number != 0:
+			return
+		if not self.editor_tab_save_button.is_sensitive():
+			return
+		if not gui_utilities.show_dialog_yes_no('Changes not saved', self.application.get_active_window(), 'Do you want to save your changes?'):
+			return
+
+		self._save_editor_file()
+
+	def signal_edit_local_file(self, _):
```

Comment:
This looks like it could be combined with the one below it. You'd need to add the additional `except IoError` block, but the rest should be fine. Once again this requires standardizing the `read_file` interface from `(ftp_)read_file`. You can pass in an additional argument via the signal handler. Connect it twice and connect the appropriate directories, local for the local one etc.